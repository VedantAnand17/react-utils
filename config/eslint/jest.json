{
  "env": {
    "browser": true,
    "jest": true
  },
  "plugins": [
    "jest"
  ],
  "rules": {
    "@typescript-eslint/no-implied-eval": 0,
    "global-require": 0,
    "import/no-dynamic-require": 0,
    "import/no-extraneous-dependencies": 0,
    "no-console": 0,
    "no-new-func": 0,
    "react/require-default-props": 0
  }
}