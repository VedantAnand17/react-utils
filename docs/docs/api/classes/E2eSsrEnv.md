# E2eSsrEnv
```jsx
import E2eSsrEnv from '@dr.pogodin/react-utils/jest-e2e-ssr-env';

// Alternatively, within an actual Jest test:
/** @jsdom-environment @dr.pogodin/react-utils/jest-e2e-ssr-env */
```
The [E2eSsrEnv] class implements a
[Jest test environment](https://jestjs.io/docs/configuration#testenvironment-string)
for end-to-end testing of SSR and client-side execution of Webpack-built code.
It should be considered as a part of [Jest utilities][JU] provided by the library,
but due to technical reasons it cannot be exported directly from [JU] sub-module,
and should stay as a separate library export.

Given a source ReactJS code, a Webpack config factory, and additional options,
it bundles the application code with Webpack, emitting output assets into
in-memory filesystem, it then generates server-side (SSR) markup for
the specified request, and provides client-side ([JsDom]) environment
for Jest testing with outcomes of Webpack build and SSR exposed via
global variables.

The environment and its options are configured in a Jest test via pragmas
in the opening docblock, _e.g._:

```jsx
/**
 * @jest-environment @dr.pogodin/react-utils/jest-e2e-ssr-env
 * @webpack-config-factory @dr.pogodin/react-utils/config/webpack/app-production.js
 * @webpack-config-options { "entry": "./path/to/webpack-entry.js" }
 * @ssr-options {
 *  "babelEnv": "production",
 *  "entry": "./path/to/Scene.jsx" }
 */
```

[Real-life examples of using this test environment](https://github.com/birdofpreyru/react-utils/tree/master/__tests__/config/stylename-generation)
(these examples are within react-utils library codebase, because of this they
import the environment with the pragma
`@jest-environment ./src/shared/utils/jest/E2eSsrEnv.js` using a local path).

## Configuration
The following docblock pragma parameters are recognized by the environment:
- `@webpack-config-factory PATH` - The path to [Webpack config factory] to
  use for the build. If `PATH` is relative, it is resolved relative to the root
  directory of Jest configuration.

- `@webpack-config-options OBJECT` - The options object to pass into
  [Webpack config factory]. The environment will automatically set this object's
  `.context` field to the current test's folder, and a new [memfs] in-memory
  filesystem will be created and provided via `.fs` option.
  You want to set `.entry` field to point to the tested app entry
  point (which should include the mounting of the app into DOM, _e.g._ by
  calling to the [client()] function).

  When [development config factory](/docs/api/configs/webpack#app-dev) is used
  you probably want to set **true** `.dontUseHmr` and
  `.dontUseReactGlobalStateDebugging` flags to avoid polluting test logs with
  logging from React Global State, and avoid problems with HMR incompatible
  with Jest testing setup without extra efforts.

  :::info CHANGELOG
  In the library versions **v1.17.2** &div; **v1.17.4** `.dontEmitBuildInfo`
  flag was automatically set **true**. Since **v1.17.5** the `.fs` option is
  used instead to collect in a virtual filesystem artifacts created during
  the Webpack config generation.
  :::

- `@ssr-options OBJECT` - The options for SSR rendering. The object is passed
  into the standard [server()] factory, but there are a few special fields
  accepted within this object by the environment:
  - `.babelEnv` - **string** - Optional. Allows to enforce the correct Babel
    environment for SSR in the test. _e.g._ "**development**" or "**production**".
  - `.buildInfo` - **object** - Optional. Allows to enforce specific
    "[build info]" data to be used in SSR.
  - `.entry` - **string** - The path to app entry to use in SSR. You want to
    have it different from Webpack's entry point, as it should not include
    the app mounting into DOM ([client()] call). Typically, you want the entry
    point for Webpack to be a thin wrapper module around this entry point for SSR.

    The environment will use this path to automatically set `.Application`
    field to the root application component to render.

    If relative path is provided, it is resolved relative to the test folder.
  - `.entryExportName` - **string** - By default `.Application` option is set
    based on the "**default**" export from the module pointed by `.entry` path.
    When this field is set, the corresponding named export is used instead.

  - `.root` - **string** - By default, SSR runs with the current working
    directory, `process.cwd()`,
    as the [Babel's root](https://babeljs.io/docs/en/options#root),
    where default configuration files for Babel are looked up.
    This option allows to alter it. It currently accepts a single valid value:
    - "`TEST`" - To use the current test folder as the root.

      :::tip TIP
      You may want to use this option to place a custom `babel.config.json` file
      into the test folder. Mind that to completely override Babel settings for
      the tested codebase, you probably should set to **false**
      the [babelrc](https://babeljs.io/docs/en/options#babelrc) flag
      in that  `babel.config.json`  config.
      :::

- `@ssr-request OBJECT` - Optional. The request object to use for SSR,
  _e.g._ to provide a custom `url`.

- `@no-ssr true` - Set this to opt out of SSR, if only Webpack build is
  needed for the test.

## Outputs
The environment exposes outcomes of the Webpack build and SSR rendering to
the actual test runtime via the following global variables:
- `buildInfo` - **object** - The "[build info]" object from the Webpack build.
- `dom` - **object** - The instance of [JsDom] object. For example, you may
  use its `.reconfigure()` method to change URL being used by the client-side.
- `REACT_UTILS_FORCE_CLIENT_SIDE` - **boolean** - This global flag is
  automatically set **true** to force-emulate the client-side execution
  for React Utils library within the environment.
- `ssrMarkup` - **string** - HTML markup generated by SSR.
- `ssrOptions` - **object** - Options used by SSR.
- `ssrStatus` - **number** - HTTP status code generated during SSR rendering.
- `webpackConfig` - **object** - The config generated by the provided config
  factory and used for the Webpack build.
- `webpackOutputFs` - **object** - The in-memory filesystem ([memfs] instance)
  with outputs from the Webpack build.
- `webpackStats` - **object** - [Webpack stats](https://webpack.js.org/api/stats/)
  from the Webpack build, converted to JSON format.

## Hints
Within the Jest test using [E2eSsrEnv] you may use the following code patterns
to test SSR and Webpack-build outcomes.

```jsx
/**
 * @jest-environment @dr.pogodin/react-utils/jest-e2e-ssr-env
 * @webpack-config-factory @dr.pogodin/react-utils/config/webpack/app-development.js
 * @webpack-config-options {
 *  "entry": "./path/to/webpack-entry.js",
 *  "dontUseHmr": true,
 *  "dontUseReactGlobalStateDebugging": true }
 * @ssr-options { "entry": "./path/to/Scene.jsx" }
 */
// Above is a sample docblock comment configuring the environment with
// our standard development config for Webpack.

// This is needed to emulate client-side hydration further below.
import { act } from 'react-dom/test-utils';

// The path for Webpack outputs within the in-memory filesystem.
const outputPath = global.webpackConfig.output.path;

// Information on Webpack outputs by bundle chunk names.
const { namedChunkGroups } = global.webpackStats;

// Listing the content of Webpack output folder.
const fs = global.webpackOutputFs;
console.log(fs.readdirSync(outputPath));

// Mounting the SSR-generated markup into DOM.
// You may need document.open() and document.close() calls if you are
// to remount the document.
document.write(global.ssrMarkup);

// This is the container of the main application content,
// if your rely on the standard client() function from the library.
const container = document.querySelector('#react-view');

// To emulate the client-side execution (hydration), you need to do
// the following within a test. The following example assumes we use
// the standard development config factory for the Webpack build,
// which guarantees the main JS bundle file is called "main.js"
// (in case of production config the filename should be found
// within Webpack build stats).
it('sample test', async ()= > {
  // This is the main app content originating from the SSR.
  const markup = container.innerHTML;

  // This does the client-side hydration.
  const js = fs.readFileSync(`${outputPath}/main.js`, 'utf8');
  await act(new Function(js)); // eslint-disable-line no-new-func

  // This checks the outcome of the client-side hydration matches the original
  // SSR markup. Note: if hydration fails, the React automatically prints out
  // the error message, but that message is not automatically caught and reported
  // by the test. If you need to fail the test on hydration failure, you should
  // do the following check.
  expect(document.querySelector('#react-view').innerHTML).toBe(markup);
});
```

<!-- links -->
[build info]: /docs/api/configs/webpack#build-info
[client()]: /docs/api/functions/client
[E2eSsrEnv]: /docs/api/classes/E2eSsrEnv
[JsDom]: https://github.com/jsdom/jsdom
[JU]: /docs/api/utils/jest-utils
[memfs]: https://www.npmjs.com/package/memfs
[server()]: /docs/api/functions/server
[Webpack config factory]: /docs/api/configs/webpack
