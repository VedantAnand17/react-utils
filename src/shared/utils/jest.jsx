/**
 * Utilities for Jest tests.
 */

/* NOTE: At the moment here is no intention to use this module for runtime
 * execution, only for tests; thus we depend on Jest environment and dev
 * dependencies. */

/* TODO: Revise.  */

/* global expect, jest, document */
/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';

import mockdate from 'mockdate';
import PT from 'prop-types';
import { render as newRender, unmountComponentAtNode } from 'react-dom';
import TU from 'react-dom/test-utils';

/* eslint-disable import/no-extraneous-dependencies */
import Renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
/* eslint-enable import/no-extraneous-dependencies */

/* NEW TESTING METHODS */

export { act } from 'react-dom/test-utils';

const originalProcessVersions = process.versions;

export function mockClientSide() {
  Object.defineProperty(process, 'versions', { value: null });
}

export function unmockClientSide() {
  Object.defineProperty(process, 'versions', {
    value: originalProcessVersions,
    writable: false,
  });
}

/**
 * Generates a mock UUID.
 * @param {Number} seed
 * @return {String}
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param {Number} time Time step [ms].
 * @returns {Promise} Wait for this to "jump after" any async code which should
 *  be executed because of the mock time movement.
 */
export async function mockTimer(time) {
  mockdate.set(time + Date.now());
  jest.advanceTimersByTime(time);
}

/**
 * Mounts `Scene` to the DOM, and returns the root scene element.
 * @param {React.ReactNode} scene
 * @return {HTMLElement}
 */
export function mount(scene) {
  const res = document.createElement('div');
  document.body.appendChild(res);
  newRender(scene, res);
  return res;
}

/**
 * Unmounts `scene` from the DOM.
 * @param {HTMLElement} scene
 */
export function unmount(scene) {
  unmountComponentAtNode(scene);
  scene.remove();
}

/* OLD STUFF BELOW THIS MARK */

/**
 * Just an alias for TU.findRenderedDOMComponentWithClass(..).
 * @param {Object} dom
 * @param {String} className
 * @return {Object}
 */
export function findInDomByClass(dom, className) {
  return TU.findRenderedDOMComponentWithClass(dom, className);
}

/**
 * Just an alias for TU.scryRenderedDOMComponentsWithClass(..).
 * @param {Object} dom
 * @param {Stirng} className
 * @return {Array}
 */
export function findInDomManyByClass(dom, className) {
  return TU.scryRenderedDOMComponentsWithClass(dom, className);
}

/**
 * Auxiliary wrapper around ReactJS Test Renderer.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of the rendered tree.
 */
export function render(component) {
  return Renderer.create(component).toJSON();
}

/* The Wrapper is necessary for the "renderDom(..)" function, because
 * the "renderIntoDocument(..)" function from "react-dom/test-utils" works
 * only with state components, so we have to wrap our ReactJS components into
 * such Wrapper. */
class Wrapper extends React.Component {
  componentDidMount() {}

  render() {
    const { children } = this.props;
    return children;
  }
}

Wrapper.propTypes = {
  children: PT.node.isRequired,
};

/**
 * Renders given component into DOM, using "react-dom/test-utils".
 * @param {Object} component ReactJS component to render.
 * @return {Object} Rendered DOM.
 */
export function renderDom(component) {
  return TU.renderIntoDocument((
    <Wrapper>
      {component}
    </Wrapper>
  ));
}

/**
 * Auxiliary wrapper around ReactJS Shallow Test Renderer.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of the shallow component's render tree.
 */
export function shallowRender(component) {
  const renderer = new ShallowRenderer();
  renderer.render(component);
  return renderer.getRenderOutput();
}

/**
 * Renders provided ReactJS component with ReactJS Shallow Test Renderer,
 * creates/tests its snapshot, and returns the render result.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of shallow render.
 */
export function shallowSnapshot(component) {
  const res = shallowRender(component);
  expect(res).toMatchSnapshot();
  return res;
}

/**
 * Renders provided ReactJS component with ReactJS Test Renderer,
 * creates/tests the snapshot, and returns the render.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON render of the component.
 */
export function snapshot(component) {
  const res = render(component);
  expect(res).toMatchSnapshot();
  return res;
}

export const simulate = TU.Simulate;
