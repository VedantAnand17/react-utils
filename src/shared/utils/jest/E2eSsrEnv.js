/**
 * Jest environment for end-to-end SSR and client-side testing. It relies on
 * the standard react-utils mechanics to execute SSR of given scene, and also
 * Webpack build of the code for client-side execution, it further exposes
 * Jsdom environment for the client-side testing of the outcomes.
 */
/* eslint-disable global-require, import/no-dynamic-require */

// BEWARE: The module is not imported into the JU module / the main assembly of
// the library, because doing so easily breaks stuff:
//  1)  This module depends on Node-specific modules, which would make JU
//      incompatible with JsDom if included into JU.
//  2)  If this module is weakly imported from somewhere else in the lib,
//      it seems to randomly break tests using it for a different reason,
//      probably some sort of a require-loop, or some issues with weak
//      require in that scenario.

import path from 'path';
import ssrFactory from 'server/renderer';

import { defaults, noop, set } from 'lodash';

// As this environment is a part of the Jest testing utils,
// we assume development dependencies are available when it is used.
/* eslint-disable import/no-extraneous-dependencies */
import register from '@babel/register';
import JsdomEnv from 'jest-environment-jsdom';
import { createFsFromVolume, Volume } from 'memfs';
import webpack from 'webpack';
/* eslint-enable import/no-extraneous-dependencies */

export default class E2eSsrEnv extends JsdomEnv {
  /**
   * Loads Webpack config, and exposes it to the environment via global
   * webpackConfig object.
   */
  loadWebpackConfig() {
    let options = this.pragmas['webpack-config-options'];
    options = options ? JSON.parse(options) : {};
    defaults(options, {
      context: this.testFolder,
      fs: this.global.webpackOutputFs,
    });

    let factory = this.pragmas['webpack-config-factory'] || '';
    factory = require(path.resolve(this.rootDir, factory));
    this.global.webpackConfig = factory(options);

    const fs = this.global.webpackOutputFs;
    let buildInfo = `${options.context}/.build-info`;
    if (fs.existsSync(buildInfo)) {
      buildInfo = fs.readFileSync(buildInfo, 'utf8');
      this.global.buildInfo = JSON.parse(buildInfo);
    }
  }

  /**
   * Executes Webpack build.
   * @return {Promise}
   */
  async runWebpack() {
    this.loadWebpackConfig();

    const compiler = webpack(this.global.webpackConfig);
    compiler.outputFileSystem = this.global.webpackOutputFs;
    return new Promise((done, fail) => {
      compiler.run((err, stats) => {
        if (err) fail(err);

        this.global.webpackStats = stats.toJson();

        // Keeps reference to the raw Webpack stats object, which should be
        // explicitly passed to the server-side renderer alongside the request,
        // so that it can to pick up asset paths for different named chunks.
        this.webpackStats = stats;

        done();
      });
    });
  }

  async runSsr() {
    let options = this.pragmas['ssr-options'];
    options = options ? JSON.parse(options) : {};

    // TODO: This is temporary to shortcut the logging added to SSR.
    if (options.logger === undefined) {
      options.logger = {
        debug: noop,
        info: noop,
        log: noop,
        warn: noop,
      };
    }

    let root;
    switch (options.root) {
      case 'TEST': root = this.testFolder; break;
      default: root = process.cwd();
    }

    // Note: This enables Babel transformation for the code dynamically loaded
    // below, as the usual Jest Babel setup does not seem to apply to
    // the environment code, and imports from it.
    register({
      envName: options.babelEnv,
      extensions: ['.js', '.jsx', '.svg'],
      root,
    });

    if (!options.buildInfo) options.buildInfo = this.global.buildInfo;

    if (options.entry) {
      const p = path.resolve(this.testFolder, options.entry);
      options.Application = require(p)[options.entryExportName || 'default'];
    }

    const renderer = ssrFactory(this.global.webpackConfig, options);
    let status = 200; // OK
    const markup = await new Promise((done, fail) => {
      renderer(
        this.ssrRequest,

        // TODO: This will do for now, with the current implementation of
        // the renderer, but it will require a rework once the renderer is
        // updated to do streaming.
        {
          cookie: noop,
          send: done,
          set: noop,
          status: (value) => {
            status = value;
          },

          // This is how up-to-date Webpack stats are passed to the server in
          // development mode, and we use this here always, instead of having
          // to pass some information via filesystem.
          locals: {
            webpack: {
              devMiddleware: {
                stats: this.webpackStats,
              },
            },
          },
        },

        (error) => {
          if (error) fail(error);
          else done('');
        },
      );
    });

    this.global.ssrMarkup = markup;
    this.global.ssrOptions = options;
    this.global.ssrStatus = status;
  }

  constructor(config, context) {
    const pragmas = context.docblockPragmas;
    let request = pragmas['ssr-request'];
    request = request ? JSON.parse(request) : {};
    if (!request.url) request.url = '/';
    request.csrfToken = noop;

    // This ensures the initial JsDom URL matches the value we use for SSR.
    set(
      config.projectConfig,
      'testEnvironmentOptions.url',
      `http://localhost${request.url}`,
    );

    super(config, context);

    this.global.dom = this.dom;
    this.global.webpackOutputFs = createFsFromVolume(new Volume());

    // Extracts necessary settings from config and context.
    const { projectConfig } = config;
    this.rootDir = projectConfig.rootDir;
    this.testFolder = path.dirname(context.testPath);
    this.withSsr = !pragmas['no-ssr'];
    this.ssrRequest = request;
    this.pragmas = pragmas;
  }

  async setup() {
    await super.setup();
    await this.runWebpack();
    if (this.withSsr) await this.runSsr();
    this.global.REACT_UTILS_FORCE_CLIENT_SIDE = true;
  }

  async teardown() {
    delete this.global.REACT_UTILS_FORCE_CLIENT_SIDE;

    // Resets module cache and @babel/register. Effectively this ensures that
    // the next time an instance of this environment is set up, all modules are
    // transformed by Babel from scratch, thus taking into account the latest
    // Babel config (which may change between different environment instances,
    // which does not seem to be taken into account by Babel / Node caches
    // automatically).
    Object.keys(require.cache).forEach((key) => {
      delete require.cache[key];
    });
    register.revert();
    super.teardown();
  }
}
