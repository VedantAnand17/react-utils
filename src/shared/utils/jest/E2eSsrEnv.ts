/**
 * Jest environment for end-to-end SSR and client-side testing. It relies on
 * the standard react-utils mechanics to execute SSR of given scene, and also
 * Webpack build of the code for client-side execution, it further exposes
 * Jsdom environment for the client-side testing of the outcomes.
 */
/* eslint-disable global-require, import/no-dynamic-require */

// BEWARE: The module is not imported into the JU module / the main assembly of
// the library, because doing so easily breaks stuff:
//  1)  This module depends on Node-specific modules, which would make JU
//      incompatible with JsDom if included into JU.
//  2)  If this module is weakly imported from somewhere else in the lib,
//      it seems to randomly break tests using it for a different reason,
//      probably some sort of a require-loop, or some issues with weak
//      require in that scenario.

import path from 'path';

import type { Request, Response } from 'express';
import { defaults, noop, set } from 'lodash';

// As this environment is a part of the Jest testing utils,
// we assume development dependencies are available when it is used.
/* eslint-disable import/no-extraneous-dependencies */
import register from '@babel/register/experimental-worker';

import JsdomEnv from 'jest-environment-jsdom';
import { type IFs, createFsFromVolume, Volume } from 'memfs';
import webpack from 'webpack';
/* eslint-enable import/no-extraneous-dependencies */

import ssrFactory from 'server/renderer';

import type {
  EnvironmentContext,
  JestEnvironmentConfig,
} from '@jest/environment';

export default class E2eSsrEnv extends JsdomEnv {
  pragmas: Record<string, string | string[]>;

  ssrRequest: object;

  rootDir: string;

  testFolder: string;

  withSsr: boolean;

  webpackStats?: webpack.StatsCompilation;

  /**
   * Loads Webpack config, and exposes it to the environment via global
   * webpackConfig object.
   */
  loadWebpackConfig() {
    const optionsString = this.pragmas['webpack-config-options'] as string;

    const options = (optionsString
      ? JSON.parse(optionsString) : {}) as webpack.Configuration;

    defaults(options, {
      context: this.testFolder,
      fs: this.global.webpackOutputFs,
    });

    const factoryPath = this.pragmas['webpack-config-factory'] as string;
    let factory = require(path.resolve(this.rootDir, factoryPath));
    factory = 'default' in factory ? factory.default : factory;

    this.global.webpackConfig = factory(options);

    const fs = this.global.webpackOutputFs as IFs;
    let buildInfo = `${options.context}/.build-info`;
    if (fs.existsSync(buildInfo)) {
      buildInfo = fs.readFileSync(buildInfo, 'utf8') as string;
      this.global.buildInfo = JSON.parse(buildInfo);
    }
  }

  /**
   * Executes Webpack build.
   * @return {Promise}
   */
  async runWebpack() {
    this.loadWebpackConfig();

    const compiler = webpack(this.global.webpackConfig as webpack.Configuration);
    compiler.outputFileSystem = this.global.webpackOutputFs;
    return new Promise<void>((done, fail) => {
      compiler.run((err, stats) => {
        if (err) fail(err);
        if (stats?.hasErrors()) {
          // eslint-disable-next-line no-console
          console.error(stats.toJson().errors);
          fail(Error('Webpack compilation failed'));
        }

        this.global.webpackStats = stats?.toJson();

        // Keeps reference to the raw Webpack stats object, which should be
        // explicitly passed to the server-side renderer alongside the request,
        // so that it can to pick up asset paths for different named chunks.
        this.webpackStats = stats;

        done();
      });
    });
  }

  async runSsr() {
    const optionsString = this.pragmas['ssr-options'] as string;
    const options = optionsString ? JSON.parse(optionsString) : {};

    // TODO: This is temporary to shortcut the logging added to SSR.
    if (options.logger === undefined) {
      options.logger = {
        debug: noop,
        info: noop,
        log: noop,
        warn: noop,
      };
    }

    if (!options.buildInfo) options.buildInfo = this.global.buildInfo;

    let cleanup: (() => void) | undefined;

    if (options.entry) {
      const p = path.resolve(this.testFolder, options.entry);
      const module = require(p);
      cleanup = module.cleanup;
      options.Application = module[options.entryExportName || 'default'];
    }

    const renderer = ssrFactory(this.global.webpackConfig!, options);
    let status = 200; // OK
    const markup = await new Promise<string>((done, fail) => {
      renderer(
        this.ssrRequest as Request,

        // TODO: This will do for now, with the current implementation of
        // the renderer, but it will require a rework once the renderer is
        // updated to do streaming.
        ({
          cookie: noop,
          send: done,
          set: noop,
          status: (value: number) => {
            status = value;
          },

          // This is how up-to-date Webpack stats are passed to the server in
          // development mode, and we use this here always, instead of having
          // to pass some information via filesystem.
          locals: {
            webpack: {
              devMiddleware: {
                stats: this.webpackStats,
              },
            },
          },
        } as unknown) as Response,

        (error) => {
          if (error) fail(error);
          else done('');
        },
      );
    });

    this.global.ssrMarkup = markup;
    this.global.ssrOptions = options;
    this.global.ssrStatus = status;

    if (cleanup) cleanup();
  }

  constructor(
    config: JestEnvironmentConfig,
    context: EnvironmentContext,
  ) {
    const pragmas = context.docblockPragmas;

    const requestString = pragmas['ssr-request'] as string;
    const request = requestString ? JSON.parse(requestString) : {};

    if (!request.url) request.url = '/';
    request.csrfToken = noop;

    // This ensures the initial JsDom URL matches the value we use for SSR.
    set(
      config.projectConfig,
      'testEnvironmentOptions.url',
      `http://localhost${request.url}`,
    );

    super(config, context);

    this.global.dom = this.dom;
    this.global.webpackOutputFs = createFsFromVolume(new Volume());

    // Extracts necessary settings from config and context.
    const { projectConfig } = config;
    this.rootDir = projectConfig.rootDir;
    this.testFolder = path.dirname(context.testPath);
    this.withSsr = !pragmas['no-ssr'];
    this.ssrRequest = request;
    this.pragmas = pragmas;

    // The usual "babel-jest" transformation setup does not apply to
    // the environment code and imports from it, this workaround enables it.
    const optionsString = this.pragmas['ssr-options'] as string;
    const options = optionsString ? JSON.parse(optionsString) : {};
    let root;
    switch (options.root) {
      case 'TEST': root = this.testFolder; break;
      default: root = process.cwd();
    }
    register({
      envName: options.babelEnv,
      extensions: ['.js', '.jsx', '.ts', '.tsx', '.svg'],
      root,
    });
  }

  async setup() {
    await super.setup();
    await this.runWebpack();
    if (this.withSsr) await this.runSsr();
    this.global.REACT_UTILS_FORCE_CLIENT_SIDE = true;
  }

  async teardown() {
    delete this.global.REACT_UTILS_FORCE_CLIENT_SIDE;

    // Resets module cache and @babel/register. Effectively this ensures that
    // the next time an instance of this environment is set up, all modules are
    // transformed by Babel from scratch, thus taking into account the latest
    // Babel config (which may change between different environment instances,
    // which does not seem to be taken into account by Babel / Node caches
    // automatically).
    Object.keys(require.cache).forEach((key) => {
      delete require.cache[key];
    });
    register.revert();
    super.teardown();
  }
}
