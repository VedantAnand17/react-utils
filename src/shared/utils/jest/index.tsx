/* global expect, jest, document */
/* eslint-disable import/no-extraneous-dependencies */

import mockdate from 'mockdate';
import { type ReactElement, type ReactNode } from 'react';
import { type Root, createRoot } from 'react-dom/client';
import TU, { act } from 'react-dom/test-utils';

/* eslint-disable import/no-extraneous-dependencies */
import Renderer from 'react-test-renderer';
import { type ShallowRenderer, createRenderer } from 'react-test-renderer/shallow';
/* eslint-enable import/no-extraneous-dependencies */

/**
 * An alias for [act(..)](https://reactjs.org/docs/test-utils.html#act)
 * from `react-dom/test-utils`.
 * @param {function} action
 */
export { act };

declare global {
  // eslint-disable-next-line no-var, vars-on-top
  var IS_REACT_ACT_ENVIRONMENT: boolean | undefined;
}

global.IS_REACT_ACT_ENVIRONMENT = true;

const originalProcessVersions = process.versions;

/**
 * Tricks **react-utils** into thinking the test is running within client-side
 * (browser) environment.
 */
export function mockClientSide() {
  Object.defineProperty(process, 'versions', { value: null });
}

/**
 * Reverts the effect of {@link module:JU.mockClientSide mockClientSide(..)}.
 */
export function unmockClientSide() {
  Object.defineProperty(process, 'versions', {
    value: originalProcessVersions,
    writable: false,
  });
}

/**
 * Generates a mock UUID, or better said it determenistically transforms given
 * `seed` number into a UUID-formatted string.
 * @param {number} seed
 * @return {string}
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param {number} time Time step [ms].
 * @returns {Promise} Wait for this to "jump after" any async code which should
 * be executed because of the mock time movement.
 */
export async function mockTimer(time: number) {
  mockdate.set(time + Date.now());
  jest.advanceTimersByTime(time);
}

type MountedSceneT = HTMLElement & {
  destroy: () => void;
};

/**
 * Mounts `scene` to the DOM, and returns the root scene element.
 * @param {React.ReactNode} scene
 * @return Created container DOM element with destroy() function
 *  attached.
 */
export function mount(scene: ReactNode): MountedSceneT {
  let root: Root;
  const element = document.createElement('div');
  document.body.appendChild(element);

  const res: MountedSceneT = (element as unknown) as MountedSceneT;
  res.destroy = () => {
    act(() => root.unmount());
    res.remove();
  };

  act(() => {
    root = createRoot(res);
    root.render(scene);
  });
  return res;
}

/* OLD STUFF BELOW THIS MARK */

/**
 * Renders provided ReactJS component into JSON representation of the component
 * tree, using [`react-test-renderer`](https://www.npmjs.com/package/react-test-renderer).
 * @param {object} component ReactJS component to render.
 * @return {object} JSON representation of the rendered tree.
 * @example
 * import { JU } from '@dr.pogodin/react-utils';
 * console.log(JU.render(<div>Example</div>));
 */
export function render(component: ReactElement) {
  return Renderer.create(component).toJSON();
}

/**
 * Generates a shallow render of given ReactJS component, using
 * [react-test-renderer/shallow](https://reactjs.org/docs/shallow-renderer.html)
 * and returns the result.
 * @param {object} component ReactJS component to render.
 * @return {object} JSON representation of the shallow component's render tree.
 */
export function shallowRender(component: ReactElement) {
  const renderer: ShallowRenderer = createRenderer();
  renderer.render(component);
  return renderer.getRenderOutput();
}

/**
 * Makes a shallow snapshot test of the given ReactJS component, and also
 * returns JSON representation of the rendered component tree. Under the hood
 * it uses {@link module:JU.shallowRender shallowRender(..)} to generate
 * the render, then executes `expect(RENDER_RESULT).toMatchSnapshot()`,
 * and finally returns the `RENDER_RESULT` to the caller.
 * @param {object} component ReactJS component to render.
 * @return {object} JSON representation of shallow render.
 */
export function shallowSnapshot(component: ReactElement) {
  const res = shallowRender(component);
  expect(res).toMatchSnapshot();
  return res;
}

/**
 * Makes snapshot test of the given ReactJS component, and also returns JSON
 * representation of the rendered component tree. Under the hood, it uses
 * {@link module:JU.render render(..)} to render it, then executes
 * `expect(RENDER_RESULT).toMatchSnapshot()`, and then returns `RENDER_RESULT`.
 * @param {object} component ReactJS component to render.
 * @return {object} JSON render of the component.
 */
export function snapshot(component: ReactElement) {
  const res = render(component);
  expect(res).toMatchSnapshot();
  return res;
}

export const simulate = TU.Simulate;
