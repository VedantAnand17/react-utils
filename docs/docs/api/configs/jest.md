# Jest
Default configuration for [Jest] testing framework.

:::caution Beware
It does strong assumptions about the host project's code structure: the host
[Jest] configuration in the `/config/jest` folder of host code; the source code
in `/src` folder; build outputs in `/build` folder; _etc._

For a project with different structure it might be easier to compose
the appropriate Jest config from scratch.
:::

To use this default configuration create two files in the `/config/jest` folder
of host project:

```js title="/config/jest/config.js"
const config = require('@dr.pogodin/react-utils/config/jest/default');

// You can tune config as necessary here.

module.exports = config;
```

```js title="/config/jest/setup.js"
import '@dr.pogodin/react-utils/config/jest/setup';
```

After that point to that host config file from the Jest execution command,
_e.g._:
```bash
jest --config config/jest/config.js
```

[Jest]: https://jestjs.io
