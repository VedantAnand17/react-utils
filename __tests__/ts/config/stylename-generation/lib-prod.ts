/**
 * @jest-environment ./src/shared/utils/jest/E2eSsrEnv.js
 * @webpack-config-factory ./config/webpack/lib-production.js
 * @webpack-config-options { "entry": "./__assets__/index.js" }
 * @no-ssr true
 */

import { global } from 'src/shared/utils/jest/E2eSsrEnv';

const fs = global.webpackOutputFs;
const outputPath = global.webpackConfig.output.path;

it('emits expected CSS', () => {
  const css = fs?.readFileSync(`${outputPath}/style.css`, 'utf8');
  expect(css).toMatchSnapshot();
});
