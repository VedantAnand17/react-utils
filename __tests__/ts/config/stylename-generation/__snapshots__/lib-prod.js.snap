// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emits expected CSS 1`] = `
".SjBfrO{background:red;height:10px;width:10px}.y4brFZ{color:blue}.MzIKg7{background:green;height:10px;width:10px}
/*# sourceMappingURL=style.css.map*/"
`;
